### PR TITLE
chore: use forwardRef type inferring

### PR DESCRIPTION
**Pull request by zheeeng**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [x] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Refactor the usages of React.forwardRef and the code style of mounting the static components on the exported component |
| 🇨🇳 Chinese | 统一 React.forwardRef 与挂载组件到主组件静态属性上的写法 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 8b7d6a3</samp>

The pull request refactors and enhances several components in the `ant-design/ant-design` repository by enabling ref forwarding and improving type safety. Ref forwarding allows users to access the internal elements or components of the wrapped components through the ref prop. Type safety is improved by using generic types, `Object.assign`, and removing unused code and imports. The components affected by the pull request are `AutoComplete`, `Avatar`, `Button`, `Cascader`, `Radio`, `Switch`, `Tag`, `Tree`, `Upload`, `BreadcrumbSeparator`, `HookModal`, and `Text`.

### <samp>🤖 Generated by Copilot at 8b7d6a3</samp>

*  Wrap components with `React.forwardRef` to expose internal refs ([link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL57-R57), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-a9110968bd15226a58d0270285b7d34c0c37c415a6e5ca231c84fb3fb82ef9efL43-R43), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-76ec2a6b6b148e569109c57f12c518ff039d0570a2c49c5055e5115f0f793e4dL4-R21), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L102-R284), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-4737f6ef564486c449ec019f2826176d8b8d186088e5e912ad06ad1919b2d804L129-R129), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-df96e61a2d0677edf2aacbeceacb0ab89df52afd4703477326af012660ac2bfeL18-R72), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-4befd0a047da31184c43bc6fef846467a6e774ecef3cf332ac25310b3d9d4a8cL15-R15), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-e6cc84da9314e7733f7658f68135d16dbc896ce22ec302e43189778cc134cf03L39-R39), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-bcceb016a2b5f77c9c373d92bb6e81d64f26323154b13adfc6b74243e78b4913L30-R30), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-abd41216def304abcca63211e726233e8c9e736775dee6d0cd4df926d981ccdfL46-R210), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-3e644abe07ccd8d5cf9e0bc6a5dd7ff5d919edc4a194b1241e9c7334b89837e5L13-R37), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-a2e7152a9f97730a600f9bda0bf15029cce0148818a6ee8d2a037fafc101427dL28-R28), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-facbb1dee1bb7771145b37f482a70fb1fe9e43942520be5a31e6a36aab1f059dL19-R19))
*  Assign components to variables using `Object.assign` instead of type assertion to preserve type information and properties ([link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL164-R163), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL175-R186), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-a9110968bd15226a58d0270285b7d34c0c37c415a6e5ca231c84fb3fb82ef9efL239-R238), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-eafdeaa24b5e2771367e4ea1383becc626c63072ccd786648d000fa1df282fcaL10-R9), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-4737f6ef564486c449ec019f2826176d8b8d186088e5e912ad06ad1919b2d804L334-R346), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-5b01f8d003cac1ca969288f2a490769e23189c2d900c18a647569bc497c2137cL18-R22), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-4befd0a047da31184c43bc6fef846467a6e774ecef3cf332ac25310b3d9d4a8cL86-R88), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-e6cc84da9314e7733f7658f68135d16dbc896ce22ec302e43189778cc134cf03L107-R106), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-bcceb016a2b5f77c9c373d92bb6e81d64f26323154b13adfc6b74243e78b4913L139-R138), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-abd41216def304abcca63211e726233e8c9e736775dee6d0cd4df926d981ccdfL46-R210), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-baa7155a0b4b9fa3e7a7ec963363bf4f15b53d6d89200e1ff90729b8720d6506L27-R28), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-a2e7152a9f97730a600f9bda0bf15029cce0148818a6ee8d2a037fafc101427dL462-R464), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-facbb1dee1bb7771145b37f482a70fb1fe9e43942520be5a31e6a36aab1f059dL255-R254))
*  Remove unused type imports and declarations from `components/avatar/index.ts`, `components/button/button.tsx`, `components/radio/index.tsx` and `components/tree/DirectoryTree.tsx` ([link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-eafdeaa24b5e2771367e4ea1383becc626c63072ccd786648d000fa1df282fcaL1-L2), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L5-R5), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L71-L78), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-5b01f8d003cac1ca969288f2a490769e23189c2d900c18a647569bc497c2137cL1-R1), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-abd41216def304abcca63211e726233e8c9e736775dee6d0cd4df926d981ccdfL23-L28), [link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-baa7155a0b4b9fa3e7a7ec963363bf4f15b53d6d89200e1ff90729b8720d6506L1-R3))
*  Export `TagType` type from `components/tag/index.tsx` to allow users to use the generic type of the `Tag` component ([link](https://github.com/ant-design/ant-design/pull/42354/files?diff=unified&w=0#diff-bcceb016a2b5f77c9c373d92bb6e81d64f26323154b13adfc6b74243e78b4913L147-R146))

### CODE REVIEW

File: components/upload/Upload.tsx [hunk 1]
```diff
@@ -459,9 +459,9 @@ const InternalUpload: React.ForwardRefRenderFunction<unknown, UploadProps> = (pr
       {renderUploadList()}
     </span>,
   );
-};
+});
 
-const Upload = React.forwardRef<unknown, UploadProps>(InternalUpload);
+const Upload = Object.assign(InternalUpload, {});
```
Comment (yoyo837) [inline components/upload/Upload.tsx:464]: 这个改动的作用是？

**Comment by li-jia-nan** [comment]
没看懂为啥用 `Object.assign`

File: components/upload/Upload.tsx (hunk 1 above)
Comment (zheeeng) [reply components/upload/Upload.tsx:464]: 只是为了对齐其它处的用法。Object.assign 的作用相当于一次作了：类型断言一个组件类型及挂载其上的属性类型 + 实现上依稀赋值挂载。
之前的写法要写很多断言、重复类型、使用 typeof 反推类型，其实只要 Object.assign 就够了，类型也是 OK 的。
因为这个模式都用到了 export const XXX = Object.assign(internalXXX，fields)，虽然有的组件可以不用 Object.assign，但如果未到还需要新增要挂载的东西，要重新改变 XXX 为 internalXXX 再 export XXX = Objet.assign，其实就可以统一写好。只是 fields 提前用 {} 占位，扩展时在 {} 中添加属性即可

File: components/auto-complete/index.tsx [hunk 2]
```diff
@@ -161,31 +158,29 @@ const AutoComplete: React.ForwardRefRenderFunction<RefSelectProps, AutoCompleteP
...
 // We don't care debug panel
 /* istanbul ignore next */
-const PurePanel = genPurePanel(RefAutoComplete);
+const PurePanel = genPurePanel(InternalAutoComplete);
 
-RefAutoComplete.Option = Option;
-RefAutoComplete._InternalPanelDoNotUseOrYouWillBeFired = PurePanel;
+const AutoComplete = Object.assign(InternalAutoComplete as AutoCompleteType, {
+  Option,
+  /** @internal */
```
Comment (MadCcc) [inline components/auto-complete/index.tsx:178]: Should not be `internal`, may cause type definition loss.

Comment (zheeeng) [reply components/auto-complete/index.tsx:178]: > Should not be `internal`, may cause type definition loss.

Is `_InternalPanelDoNotUseOrYouWillBeFired` prepared for external users?

Comment (zheeeng) [reply components/auto-complete/index.tsx:178]: Is `_InternalPanelDoNotUseOrYouWillBeFired` prepared for external users?

**Review by AbdullahWins — APPROVED** [review]
good job!

**Comment by afc163** [comment]
conflicting

**Comment by afc163** [comment]
Closing for no response.

**Pull request closed without merging**